cpp/ql/src/experimental/Security/CWE/CWE-552/DangerousUseRename.ql

```diff
@@ -0,0 +1,58 @@
+/**
+ * @name Dangerous use of rename
+ * @description If the rename function did not work correctly, simply overwriting the destination file can lead to errors.
+ * @kind problem
+ * @id cpp/dangerous-use-rename
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-552
+ */
+
+import cpp
+import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+
+/** Holds if the function argument is used in opening the file for reading or writing. */
+predicate findFileForReadOrWrite(FunctionCall fc, FunctionCall fc1, string mode, int na) {
+  fc1.getTarget().hasName(["open", "fopen"]) and
+  globalValueNumber(fc1.getArgument(0)) = globalValueNumber(fc.getArgument(na)) and
+  fc1.getArgument(1).getValue() = mode and
+  fc.getASuccessor+() = fc1
+}
+
+from FunctionCall fc
+where
+  fc.getTarget().hasName("rename") and
+  exists(IfStmt ifst, Expr ec, Expr ecd, FunctionCall fc1, FunctionCall fc2 |
+    findFileForReadOrWrite(fc, fc1, "r", 0) and
+    findFileForReadOrWrite(fc, fc2, "w", 1) and
+    ec.getValue() = "0" and
```

Comment:
Could you move this down to where it's actually used? That is, to just above `ecd.(EQExpr).hasOperands(_, ec)`